ctru-rs/src/applets/swkbd.rs

```diff
@@ -675,6 +640,292 @@ impl SoftwareKeyboard {
+                    extra.learning_data,
+                    swkbd_shared_mem_ptr.add(learning_off).cast(),
+                    1,
+                )
+            };
+        }
+
+        if extra.callback.is_some() {
+            swkbd.filter_flags |= SWKBD_FILTER_CALLBACK;
+        } else {
+            swkbd.filter_flags &= !SWKBD_FILTER_CALLBACK;
+        }
+
+        // Launch swkbd
+        unsafe {
+            swkbd.__bindgen_anon_1.reserved.fill(0);
+
+            if extra.callback.is_some() {
+                self.callback_data.extra = std::ptr::addr_of_mut!(extra);
+                self.callback_data.swkbd_shared_mem_ptr = swkbd_shared_mem_ptr;
+
+                aptSetMessageCallback(
+                    Some(Self::swkbd_message_callback),
+                    std::ptr::addr_of_mut!(self.callback_data).cast(),
+                );
+            }
+
+            aptLaunchLibraryApplet(
+                APPID_SOFTWARE_KEYBOARD,
+                swkbd as *mut _ as *mut _,
+                std::mem::size_of::<SwkbdState>(),
+                swkbd_shared_mem_handle,
+            );
+
+            if extra.callback.is_some() {
+                aptSetMessageCallback(None, std::ptr::null_mut());
+            }
+
+            let _ = svcCloseHandle(swkbd_shared_mem_handle);
+        }
+
+        let button = match swkbd.result {
+            SWKBD_D1_CLICK0 | SWKBD_D2_CLICK0 => SWKBD_BUTTON_LEFT,
+            SWKBD_D2_CLICK1 => SWKBD_BUTTON_MIDDLE,
+            SWKBD_D0_CLICK | SWKBD_D1_CLICK1 | SWKBD_D2_CLICK2 => SWKBD_BUTTON_RIGHT,
+            _ => SWKBD_BUTTON_NONE,
+        };
+
+        if swkbd.text_length > 0 {
+            let text16 = unsafe {
+                widestring::Utf16Str::from_slice_unchecked(std::slice::from_raw_parts(
+                    swkbd_shared_mem_ptr.add(swkbd.text_offset as _).cast(),
+                    swkbd.text_length as _,
+                ))
+            };
+
+            *output = text16.to_string();
+        }
+
+        if swkbd.save_state_flags & (1 << 0) != 0 {
+            unsafe {
+                std::ptr::copy_nonoverlapping(
+                    swkbd_shared_mem_ptr.add(swkbd.status_offset as _).cast(),
+                    extra.status_data,
+                    1,
+                )
+            };
+        }
+
+        if swkbd.save_state_flags & (1 << 1) != 0 {
+            unsafe {
+                std::ptr::copy_nonoverlapping(
+                    swkbd_shared_mem_ptr.add(swkbd.learning_offset as _).cast(),
+                    extra.learning_data,
+                    1,
+                )
+            };
+        }
+
+        unsafe { libc::free(swkbd_shared_mem_ptr) };
+
+        button
+    }
+
+    // A reimplementation of `swkbdMessageCallback` from `libctru/source/applets/swkbd.c`.
+    // This is only needed because the original function is private to libctru, so we can't
+    // simply reuse their version
+    #[deny(unsafe_op_in_unsafe_fn)]
```

Comment:
❤️❤️❤️

I have sometimes considered if we should consider denying as a library lint but I think it would be a nontrivial amount of work so never bothered to try